File: iosPromotionSDK/PromotionSDKUI/Endow/EndowViewModel.swift
//
//  EndowViewModel.swift
//  PromotionSDK
//
//  ViewModel cho widget `PRMEndowView` — bọc `EndowStore` (tầng UI-logic dùng chung ở promotionLogic).
//  ĐỐI ỨNG Android `EndowViewModel`: findEligible và validate&apply đi qua `EndowStore` dùng chung;
//  `PromotionSDKImpl` chỉ quan sát state + render.
//
//  Dùng `PRMStoreViewModel` (bản KHÔNG router — widget không điều hướng), nên phần giữ store /
//  observe / hop main / huỷ lúc deinit không còn viết lại ở đây.
//

import Foundation
@_implementationOnly import PRMKotlinBridge

final class EndowViewModel: PRMStoreViewModel<EndowStore> {

    /// **Tắt** cầu lỗi → effect tự động của base: widget tự hiện lỗi rồi tự `consumeError`.
    ///
    /// Trước đây tắt vì lý do khác — phải giữ `errorCode` cho tới khi vòng validate kết thúc, do
    /// `validateAndApply` là fire-and-forget nên nơi gọi phải đọc nhờ dòng state chung. Lý do đó
    /// hết: store `suspend` và **trả thẳng state cuối** của lượt gọi.
    override var autoConsumesError: Bool { false }

    init(findEligibleUseCase: FindEligibleCampaignsUseCase = FindEligibleCampaignsUseCase(),
         validateUseCase: ValidateStackableDiscountsUseCase = ValidateStackableDiscountsUseCase(),
         createRedemptionUseCase: CreateRedemptionSessionUseCase = CreateRedemptionSessionUseCase()) {
        super.init(store: EndowStore(
            findEligibleCampaignsUseCase: findEligibleUseCase,
            validateStackableDiscountsUseCase: validateUseCase,
            createRedemptionSessionUseCase: createRedemptionUseCase
        ))
    }

    // ─── Observe (đối ứng Android `uiState.collect`) ────────────────────────────
    /// Quan sát state (base đã hop main). Gọi một lần khi widget attach.
    ///
    /// Gán thẳng khe `onState` của base. Trước đây phải đi vòng qua `externalOnState` vì VM còn
    /// chiếm khe đó để chạy `handleSettle` trên MỌI state — `handleSettle` đã bỏ nên khe trả về
    /// đúng một người dùng.
    func observe(_ onState: @escaping (EndowState) -> Void) {
        self.onState = onState
        onState(currentState)   // phát ngay state hiện tại, khớp hành vi replay của StateFlow
    }

    // ─── Public API (forward xuống store) ───────────────────────────────────────
    func loadInitial() { dispatch(EndowIntentLoadInitial.shared) }

    /// Cờ hiển thị widget, đọc cache (fail-open). Rule + tên cờ nằm ở store, dùng chung Android.
    func availabilityFromCache() -> Bool { store.availabilityFromCache() }

    /// Làm mới cờ từ server rồi trả giá trị thật.
    ///
    /// `.boolValue`: hàm `suspend` trả `Boolean` bên Kotlin sang Swift thành **`KotlinBoolean`** (kiểu
    /// hộp), khác hẳn `availabilityFromCache()` không suspend — cái đó ra `Bool` thẳng.
    /// Hỏng mạng → `try?` cho nil → `true`, fail-open đúng như [PromotionFeatureGate.refresh].
    func refreshAvailability() async -> Bool {
        (try? await store.refreshAvailability())?.boolValue ?? true
    }

    /// Validate + áp offers đã chọn; [completion] gọi MỘT lần khi validate xong (thành công/thất bại).
    /// Widget-state cập nhật qua [observe]; [completion] để `PromotionSDKImpl` điều hướng (pop, callback host).
    func validateAndApply(_ offers: [EligibleOffer], completion: ((EndowState) -> Void)? = nil) {
        Task { @MainActor in
            // `suspend` bên Kotlin → `async throws` bên Swift. Trả state cuối của ĐÚNG lượt này nên
            // không còn `settleCompletion`/`sawValidating`/`handleSettle` — Android bỏ y hệt.
            let state = (try? await store.validateAndApply(offers: offers)) ?? currentState
            completion?(state)
            // Đã giao lỗi cho nơi gọi → giờ mới xoá, để lỗi cũ không dính sang vòng validate sau.
            if state.errorCode != nil { consumeError() }
        }
    }

    func setApplied(_ discounts: [EndowAppliedDiscount], unavailable: Bool) {
        dispatch(EndowIntentSetApplied(discounts: discounts, unavailable: unavailable))
    }

    /// Bấm "Thanh toán" — nghiệp vụ nằm trọn ở `EndowStore.confirmRedemption` (dùng chung Android).
    ///
    /// Trước đây iOS **không có** luồng này: `createRedemption` chỉ tồn tại ở `PromotionSDKApi`
    /// (headless, host tự gọi), còn Android thì giấu trong `PromotionIntegrateManager`. Nay một đường.
    func confirmRedemption(completion: @escaping (EndowConfirmResult) -> Void) {
        Task { @MainActor in
            // `suspend` bên Kotlin — Swift nhập vào thành `async throws`. Chỉ `CancellationException`
            // mới thoát ra được (store nuốt lỗi nghiệp vụ), nhưng vẫn bắt để không nổ ở tầng host.
            let result = (try? await store.confirmRedemption())
                ?? EndowConfirmResultFailure(errorCode: PromotionErrorCodes.shared.GENERAL)
            completion(result)
        }
    }

    func markUnavailable() { dispatch(EndowIntentMarkUnavailable.shared) }
    func clearApplied() { dispatch(EndowIntentClearApplied.shared) }
    func consumeError() { dispatch(EndowIntentConsumeError.shared) }

}

File: AndroidPromotionSDK/src/main/java/com/ttcn/prm/ui/feature/endowview/EndowViewModel.kt
package com.ttcn.prm.ui.feature.endowview

import com.ttcn.prm.ui.base.PRMStoreViewModel
import com.ttcn.promotionsdk.domain.model.eligible.EligibleOffer
import com.ttcn.promotionsdk.domain.usecase.CreateRedemptionSessionUseCase
import com.ttcn.promotionsdk.domain.usecase.FindEligibleCampaignsUseCase
import com.ttcn.promotionsdk.domain.usecase.ValidateStackableDiscountsUseCase
import com.ttcn.promotionsdk.presentation.endow.EndowConfirmResult
import com.ttcn.promotionsdk.presentation.endow.EndowIntent
import com.ttcn.promotionsdk.presentation.endow.EndowState
import com.ttcn.promotionsdk.presentation.endow.EndowStore

/**
 * ViewModel của widget [PRMEndowView] — cùng khuôn với mọi màn khác: [PRMStoreViewModel] bọc store
 * dùng chung, không thêm tầng kiến trúc nào.
 *
 * **Đổi so với bản trước, ba điểm:**
 *
 * 1. **Là `ViewModel` thật** (qua [PRMStoreViewModel]), không còn là object thường tạo bằng
 *    `create(scope)` với scope gắn vào View. Trước đây store chết theo `onDetachedFromWindow`, nên
 *    host mở màn "Chọn ưu đãi" bằng `replace()` là widget detach → `viewModel = null` → bấm "Áp
 *    dụng"/"Thanh toán" trả `PRM_ERROR_GENERAL` mà **không hề gọi mạng**. iOS không dính vì
 *    `endowVM` là property của `PromotionSDKImpl`. Nay store sống theo `ViewModelStore` của host.
 *
 * 2. **Bỏ `create()`**, lấy store qua DI nội bộ như các màn khác (`promotionViewModelFactory()`).
 *    Use case tự resolve repository từ đồ thị đã init, không nơi nào phải khai tay.
 *
 * 3. **Bỏ máy trạng thái `settleCompletion`/`sawValidating`/`handleSettle`/`consumeErrorUnlessSettling`**.
 *    Bốn thứ đó tồn tại chỉ vì `EndowStore.validateAndApply` là fire-and-forget: muốn biết kết quả
 *    phải rình `isValidating` true→false trên dòng state chung — mà `StateFlow` là **conflated** nên
 *    còn phải chặn widget xoá lỗi trước khi nơi gọi đọc kịp. Nay store `suspend` và **trả thẳng
 *    state cuối**, nên cả bốn biến mất. iOS bỏ y hệt.
 *
 * Bề mặt state là **chính [EndowState] của store**, không bọc lại — xem [PRMStoreViewModel].
 */
internal class EndowViewModel(
    findEligibleCampaignsUseCase: FindEligibleCampaignsUseCase,
    validateStackableDiscountsUseCase: ValidateStackableDiscountsUseCase,
    createRedemptionSessionUseCase: CreateRedemptionSessionUseCase,
) : PRMStoreViewModel<EndowState, EndowIntent>(
    { scope ->
        EndowStore(
            findEligibleCampaignsUseCase = findEligibleCampaignsUseCase,
            validateStackableDiscountsUseCase = validateStackableDiscountsUseCase,
            createRedemptionSessionUseCase = createRedemptionSessionUseCase,
            scope = scope,
        )
    },
) {

    /** [PRMStoreViewModel.store] khai kiểu `PRMStore`; hai hàm `suspend` dưới đây là của riêng store này. */
    private val endowStore: EndowStore get() = store as EndowStore

    fun loadInitial() = dispatch(EndowIntent.LoadInitial)

    /** Cờ hiển thị widget, đọc cache (fail-open). Rule + tên cờ nằm ở store, dùng chung iOS. */
    fun availabilityFromCache(): Boolean = endowStore.availabilityFromCache()

    /** Làm mới cờ từ server rồi trả giá trị thật. */
    suspend fun refreshAvailability(): Boolean = endowStore.refreshAvailability()

    /**
     * Validate + áp ưu đãi user chọn ở màn "Chọn ưu đãi"; **trả state cuối của đúng lượt này**.
     * Đối ứng `EndowViewModel.validateAndApply(_:)` bên iOS.
     */
    suspend fun validateAndApply(offers: List<EligibleOffer>): EndowState =
        endowStore.validateAndApply(offers)

    /** Host tự validate rồi đưa kết quả vào (giữ public API `PRMEndowView.setDiscountDetails`). */
    fun setApplied(details: List<AppliedDiscount>, unavailable: Boolean = false) =
        dispatch(EndowIntent.SetApplied(details, unavailable))

    fun markUnavailable() = dispatch(EndowIntent.MarkUnavailable)

    fun clearApplied() = dispatch(EndowIntent.ClearApplied)

    fun consumeError() = dispatch(EndowIntent.ConsumeError)

    /** Bấm "Thanh toán" — nghiệp vụ nằm trọn ở [EndowStore.confirmRedemption] (dùng chung với iOS). */
    suspend fun confirmRedemption(): EndowConfirmResult = endowStore.confirmRedemption()
}
